File: community/lib/ai.ts
/**
 * AI Auto-Response Module for PocketPing Community Edition
 * Supports OpenAI, Anthropic, and Google Gemini
 */

export interface Message {
  content: string
  sender: 'visitor' | 'operator' | 'ai'
}

export interface AIProviderConfig {
  apiKey: string
  model?: string
  systemPrompt?: string
}

export const DEFAULT_MODELS = {
  openai: 'gpt-4o',
  anthropic: 'claude-sonnet-4-5-20250929',
  google: 'gemini-2.5-flash',
} as const

// ─────────────────────────────────────────────────────────────────
// OpenAI Provider
// ─────────────────────────────────────────────────────────────────

async function generateOpenAI(messages: Message[], config: AIProviderConfig): Promise<string> {
  const model = config.model || DEFAULT_MODELS.openai

  const openaiMessages: Array<{ role: string; content: string }> = []

  if (config.systemPrompt) {
    openaiMessages.push({ role: 'system', content: config.systemPrompt })
  }

  for (const msg of messages) {
    const role = msg.sender === 'visitor' ? 'user' : 'assistant'
    const lastMsg = openaiMessages[openaiMessages.length - 1]

    if (lastMsg && lastMsg.role !== 'system' && lastMsg.role === role) {
      lastMsg.content += '\n' + msg.content
    } else {
      openaiMessages.push({ role, content: msg.content })
    }
  }

  const response = await fetch('https://api.openai.com/v1/chat/completions', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${config.apiKey}`,
    },
    body: JSON.stringify({
      model,
      messages: openaiMessages,
      max_tokens: 1000,
      temperature: 0.7,
    }),
  })

  if (!response.ok) {
    const error = await response.text()
    throw new Error(`OpenAI API error: ${response.status} - ${error}`)
  }

  const data = (await response.json()) as {
    choices?: Array<{ message?: { content?: string } }>
  }
  return data.choices?.[0]?.message?.content ?? ''
}

// ─────────────────────────────────────────────────────────────────
// Anthropic Provider
// ─────────────────────────────────────────────────────────────────

async function generateAnthropic(messages: Message[], config: AIProviderConfig): Promise<string> {
  const model = config.model || DEFAULT_MODELS.anthropic

  const anthropicMessages: Array<{ role: string; content: string }> = []

  for (const msg of messages) {
    const role = msg.sender === 'visitor' ? 'user' : 'assistant'
    const lastMsg = anthropicMessages[anthropicMessages.length - 1]

    if (lastMsg && lastMsg.role === role) {
      lastMsg.content += '\n' + msg.content
    } else {
      anthropicMessages.push({ role, content: msg.content })
    }
  }

  if (anthropicMessages.length > 0 && anthropicMessages[0].role === 'assistant') {
    anthropicMessages.shift()
  }

  const response = await fetch('https://api.anthropic.com/v1/messages', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-api-key': config.apiKey,
      'anthropic-version': '2023-06-01',
    },
    body: JSON.stringify({
      model,
      max_tokens: 1000,
      system: config.systemPrompt || 'You are a helpful customer support assistant.',
      messages: anthropicMessages,
    }),
  })

  if (!response.ok) {
    const error = await response.text()
    throw new Error(`Anthropic API error: ${response.status} - ${error}`)
  }

  const data = (await response.json()) as { content?: Array<{ text?: string }> }
  return data.content?.[0]?.text ?? ''
}

// ─────────────────────────────────────────────────────────────────
// Google Gemini Provider
// ─────────────────────────────────────────────────────────────────

async function generateGemini(messages: Message[], config: AIProviderConfig): Promise<string> {
  const model = config.model || DEFAULT_MODELS.google

  const contents: Array<{ role: string; parts: Array<{ text: string }> }> = []

  for (const msg of messages) {
    const role = msg.sender === 'visitor' ? 'user' : 'model'
    const lastContent = contents[contents.length - 1]

    if (lastContent && lastContent.role === role) {
      lastContent.parts[0].text += '\n' + msg.content
    } else {
      contents.push({
        role,
        parts: [{ text: msg.content }],
      })
    }
  }

  if (contents.length > 0 && contents[0].role === 'model') {
    contents.shift()
  }

  const requestBody: Record<string, unknown> = {
    contents,
    generationConfig: {
      maxOutputTokens: 1000,
      temperature: 0.7,
    },
  }

  if (config.systemPrompt) {
    requestBody.systemInstruction = {
      parts: [{ text: config.systemPrompt }],
    }
  }

  const response = await fetch(
    `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${config.apiKey}`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(requestBody),
    }
  )

  if (!response.ok) {
    const error = await response.text()
    throw new Error(`Gemini API error: ${response.status} - ${error}`)
  }

  const data = (await response.json()) as {
    candidates?: Array<{ content?: { parts?: Array<{ text?: string }> } }>
  }
  return data.candidates?.[0]?.content?.parts?.[0]?.text ?? ''
}

// ─────────────────────────────────────────────────────────────────
// Main Export
// ─────────────────────────────────────────────────────────────────

export type AIProvider = 'openai' | 'anthropic' | 'google'

export async function generateAIResponse(
  provider: AIProvider,
  messages: Message[],
  config: AIProviderConfig
): Promise<string> {
  switch (provider) {
    case 'openai':
      return generateOpenAI(messages, config)
    case 'anthropic':
      return generateAnthropic(messages, config)
    case 'google':
      return generateGemini(messages, config)
    default:
      throw new Error(`Unknown AI provider: ${provider}`)
  }
}

File: packages/sdk-go/ai.go
package pocketping

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
)

// DefaultAISystemPrompt is the default system prompt used for AI fallback
// when none is configured.
const DefaultAISystemPrompt = "You are a helpful customer support assistant. " +
	"Be friendly, concise, and helpful. " +
	"If you don't know something, say so and offer to connect them with a human."

// DefaultAITakeoverDelay is the default delay (in seconds) before the AI takes
// over a session when no operator is online.
const DefaultAITakeoverDelay = 300

// AIProvider is the interface for AI providers used for the offline-takeover
// fallback. Implementations generate a reply from the conversation history.
type AIProvider interface {
	// Name returns the provider's unique name (e.g. "openai").
	Name() string

	// GenerateResponse generates a reply given the conversation history and an
	// optional system prompt. messages is ordered oldest-to-newest.
	GenerateResponse(ctx context.Context, messages []Message, systemPrompt string) (string, error)

	// IsAvailable reports whether the provider is reachable/usable.
	IsAvailable(ctx context.Context) bool
}

// roleForSender maps a message sender to the "user"/"assistant" role used by
// chat-completion style APIs.
func roleForSender(sender Sender) string {
	if sender == SenderVisitor {
		return "user"
	}
	return "assistant"
}

// httpClientOr returns the provided client or http.DefaultClient when nil.
func httpClientOr(c *http.Client) *http.Client {
	if c == nil {
		return http.DefaultClient
	}
	return c
}

// ─────────────────────────────────────────────────────────────────
// OpenAIProvider
// ─────────────────────────────────────────────────────────────────

// OpenAIProvider implements AIProvider using the OpenAI Chat Completions API.
type OpenAIProvider struct {
	// APIKey is the OpenAI API key (required).
	APIKey string
	// Model is the model name (default "gpt-4o-mini").
	Model string
	// BaseURL is the API base URL (default "https://api.openai.com/v1").
	// Override for testing with an httptest server.
	BaseURL string
	// HTTPClient is the HTTP client used for requests (default http.DefaultClient).
	HTTPClient *http.Client
}

// NewOpenAIProvider creates an OpenAIProvider with the given API key and defaults.
func NewOpenAIProvider(apiKey string) *OpenAIProvider {
	return &OpenAIProvider{
		APIKey:  apiKey,
		Model:   "gpt-4o-mini",
		BaseURL: "https://api.openai.com/v1",
	}
}

// Name returns the provider name.
func (p *OpenAIProvider) Name() string { return "openai" }

func (p *OpenAIProvider) model() string {
	if p.Model != "" {
		return p.Model
	}
	return "gpt-4o-mini"
}

func (p *OpenAIProvider) baseURL() string {
	if p.BaseURL != "" {
		return strings.TrimRight(p.BaseURL, "/")
	}
	return "https://api.openai.com/v1"
}

// GenerateResponse calls POST {baseURL}/chat/completions and returns
// choices[0].message.content.
func (p *OpenAIProvider) GenerateResponse(ctx context.Context, messages []Message, systemPrompt string) (string, error) {
	chatMessages := make([]map[string]string, 0, len(messages)+1)
	if systemPrompt != "" {
		chatMessages = append(chatMessages, map[string]string{
			"role":    "system",
			"content": systemPrompt,
		})
	}
	for _, msg := range messages {
		chatMessages = append(chatMessages, map[string]string{
			"role":    roleForSender(msg.Sender),
			"content": msg.Content,
		})
	}

	body, err := json.Marshal(map[string]interface{}{
		"model":       p.model(),
		"messages":    chatMessages,
		"max_tokens":  1000,
		"temperature": 0.7,
	})
	if err != nil {
		return "", err
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, p.baseURL()+"/chat/completions", bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+p.APIKey)

	resp, err := httpClientOr(p.HTTPClient).Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return "", fmt.Errorf("openai: unexpected status %d", resp.StatusCode)
	}

	var parsed struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&parsed); err != nil {
		return "", err
	}
	if len(parsed.Choices) == 0 {
		return "", nil
	}
	return parsed.Choices[0].Message.Content, nil
}

// IsAvailable performs a GET {baseURL}/models request with the Authorization header.
func (p *OpenAIProvider) IsAvailable(ctx context.Context) bool {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, p.baseURL()+"/models", nil)
	if err != nil {
		return false
	}
	req.Header.Set("Authorization", "Bearer "+p.APIKey)

	resp, err := httpClientOr(p.HTTPClient).Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()
	return resp.StatusCode >= 200 && resp.StatusCode < 300
}

var _ AIProvider = (*OpenAIProvider)(nil)

// ─────────────────────────────────────────────────────────────────
// AnthropicProvider
// ─────────────────────────────────────────────────────────────────

// AnthropicProvider implements AIProvider using the Anthropic Messages API.
type AnthropicProvider struct {
	// APIKey is the Anthropic API key (required).
	APIKey string
	// Model is the model name (default "claude-sonnet-4-20250514").
	Model string
	// BaseURL is the API base URL (default "https://api.anthropic.com/v1").
	// Override for testing with an httptest server.
	BaseURL string
	// HTTPClient is the HTTP client used for requests (default http.DefaultClient).
	HTTPClient *http.Client
}

// NewAnthropicProvider creates an AnthropicProvider with the given API key and defaults.
func NewAnthropicProvider(apiKey string) *AnthropicProvider {
	return &AnthropicProvider{
		APIKey:  apiKey,
		Model:   "claude-sonnet-4-20250514",
		BaseURL: "https://api.anthropic.com/v1",
	}
}

// Name returns the provider name.
func (p *AnthropicProvider) Name() string { return "anthropic" }

func (p *AnthropicProvider) model() string {
	if p.Model != "" {
		return p.Model
	}
	return "claude-sonnet-4-20250514"
}

func (p *AnthropicProvider) baseURL() string {
	if p.BaseURL != "" {
		return strings.TrimRight(p.BaseURL, "/")
	}
	return "https://api.anthropic.com/v1"
}

// GenerateResponse calls POST {baseURL}/messages and returns content[0].text.
func (p *AnthropicProvider) GenerateResponse(ctx context.Context, messages []Message, systemPrompt string) (string, error) {
	system := systemPrompt
	if system == "" {
		system = "You are a helpful customer support assistant."
	}

	chatMessages := make([]map[string]string, 0, len(messages))
	for _, msg := range messages {
		chatMessages = append(chatMessages, map[string]string{
			"role":    roleForSender(msg.Sender),
			"content": msg.Content,
		})
	}

	body, err := json.Marshal(map[string]interface{}{
		"model":      p.model(),
		"max_tokens": 1000,
		"system":     system,
		"messages":   chatMessages,
	})
	if err != nil {
		return "", err
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, p.baseURL()+"/messages", bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", p.APIKey)
	req.Header.Set("anthropic-version", "2023-06-01")

	resp, err := httpClientOr(p.HTTPClient).Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return "", fmt.Errorf("anthropic: unexpected status %d", resp.StatusCode)
	}

	var parsed struct {
		Content []struct {
			Text string `json:"text"`
		} `json:"content"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&parsed); err != nil {
		return "", err
	}
	if len(parsed.Content) == 0 {
		return "", nil
	}
	return parsed.Content[0].Text, nil
}

// IsAvailable returns true if an API key is set (no health endpoint).
func (p *AnthropicProvider) IsAvailable(ctx context.Context) bool {
	return p.APIKey != ""
}

var _ AIProvider = (*AnthropicProvider)(nil)

// ─────────────────────────────────────────────────────────────────
// GeminiProvider
// ─────────────────────────────────────────────────────────────────

// GeminiProvider implements AIProvider using the Google Gemini API.
type GeminiProvider struct {
	// APIKey is the Gemini API key (required).
	APIKey string
	// Model is the model name (default "gemini-1.5-flash").
	Model string
	// BaseURL is the API base host (default "https://generativelanguage.googleapis.com/v1beta").
	// Override for testing with an httptest server.
	BaseURL string
	// HTTPClient is the HTTP client used for requests (default http.DefaultClient).
	HTTPClient *http.Client
}

// NewGeminiProvider creates a GeminiProvider with the given API key and defaults.
func NewGeminiProvider(apiKey string) *GeminiProvider {
	return &GeminiProvider{
		APIKey:  apiKey,
		Model:   "gemini-1.5-flash",
		BaseURL: "https://generativelanguage.googleapis.com/v1beta",
	}
}

// Name returns the provider name.
func (p *GeminiProvider) Name() string { return "gemini" }

func (p *GeminiProvider) model() string {
	if p.Model != "" {
		return p.Model
	}
	return "gemini-1.5-flash"
}

func (p *GeminiProvider) baseURL() string {
	if p.BaseURL != "" {
		return strings.TrimRight(p.BaseURL, "/")
	}
	return "https://generativelanguage.googleapis.com/v1beta"
}

type geminiPart struct {
	Text string `json:"text"`
}

type geminiContent struct {
	Role  string       `json:"role"`
	Parts []geminiPart `json:"parts"`
}

// GenerateResponse calls POST {baseURL}/models/{model}:generateContent and
// returns candidates[0].content.parts[0].text.
func (p *GeminiProvider) GenerateResponse(ctx context.Context, messages []Message, systemPrompt string) (string, error) {
	contents := make([]geminiContent, 0, len(messages))
	for _, msg := range messages {
		role := "user"
		if msg.Sender != SenderVisitor {
			role = "model"
		}
		contents = append(contents, geminiContent{
			Role:  role,
			Parts: []geminiPart{{Text: msg.Content}},
		})
	}

	// If a system prompt is set and the first content is a user turn, prepend
	// the system prompt to its first part text.
	if systemPrompt != "" && len(contents) > 0 && contents[0].Role == "user" && len(contents[0].Parts) > 0 {
		contents[0].Parts[0].Text = systemPrompt + "\n\nUser: " + contents[0].Parts[0].Text
	}

	body, err := json.Marshal(map[string]interface{}{
		"contents": contents,
		"generationConfig": map[string]interface{}{
			"maxOutputTokens": 1000,
			"temperature":     0.7,
		},
	})
	if err != nil {
		return "", err
	}

	url := fmt.Sprintf("%s/models/%s:generateContent?key=%s", p.baseURL(), p.model(), p.APIKey)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := httpClientOr(p.HTTPClient).Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return "", fmt.Errorf("gemini: unexpected status %d", resp.StatusCode)
	}

	var parsed struct {
		Candidates []struct {
			Content struct {
				Parts []geminiPart `json:"parts"`
			} `json:"content"`
		} `json:"candidates"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&parsed); err != nil {
		return "", err
	}
	if len(parsed.Candidates) == 0 || len(parsed.Candidates[0].Content.Parts) == 0 {
		return "", nil
	}
	return parsed.Candidates[0].Content.Parts[0].Text, nil
}

// IsAvailable returns true if an API key is set.
func (p *GeminiProvider) IsAvailable(ctx context.Context) bool {
	return p.APIKey != ""
}

var _ AIProvider = (*GeminiProvider)(nil)
